src/test/osd/bad-inc-map.sh

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+source $CEPH_ROOT/qa/standalone/ceph-helpers.sh
+
+mon_port=$(get_unused_port)
+
+function run() {
+    local dir=$1
+    shift
+
+    export CEPH_MON="127.0.0.1:$mon_port"
+    export CEPH_ARGS
+    CEPH_ARGS+="--fsid=$(uuidgen) --auth-supported=none "
+    CEPH_ARGS+="--mon-host=$CEPH_MON "
+    set -e
+
+    local funcs=${@:-$(set | sed -n -e 's/^\(TEST_[0-9a-z_]*\) .*/\1/p')}
+    for func in $funcs ; do
+        setup $dir || return 1
+	$func $dir || return 1
+        teardown $dir || return 1
+    done
+}
+
+function TEST_bad_inc_map() {
+    local dir=$1
+
+    run_mon $dir a
+    run_mgr $dir x
+    TIMEOUT=60 run_osd $dir 0 --osd-inject-bad-map-crc-probability 1 || return 1
```

Comment:
Does this really detect of the osd crashes?

Reply:
I confirmed earlier today make check failed.

Reply:
(The osd doesn't even start the first time if osd-inject-bad-map-crc-probability is 1)

Reply:
Here's the failed make check: https://jenkins.ceph.com/job/ceph-pull-requests/56382/

Reply:
Sorry, my mistake, this test is supposed to show that the OSD doesn't crash with your change. RIght?  But during startup the OSD is unable to initialize itself if it can't handle any maps.

Reply:
That's right -- with the fix here then the injected crc failure will trigger `request_full_map(e, last)` then bail out from handle_osd_maps before committing the bogus transaction which would later crash in _committed_osd_maps. The OSD will then subsequently get a full osdmap and start up correctly.

Reply:
You need to set that configuration after the OSD has started (run_osd), then do something to trigger a map change so that handle_osd_map() is called.  You could also check the osd log for the new dout() you have.

I would think that the OSD would silently ignore whatever map change occurred.  So isn't an assert() the real answer here?  Or can the map be rehandled as a non-incremental?  Is the ceph_assert(first <= last); supposed to catch if the map wasn't handled?

Reply:
Could the claims in https://github.com/ceph/ceph/pull/35218 not be correct?

Reply:
> You need to set that configuration after the OSD has started (run_osd), then do something to trigger a map change so that handle_osd_map() is called. You could also check the osd log for the new dout() you have.

yes the test could be improved. I'm running out of time here today so can aim for that tomorrow.

> I would think that the OSD would silently ignore whatever map change occurred. So isn't an assert() the real answer here? Or can the map be rehandled as a non-incremental? Is the ceph_assert(first <= last); supposed to catch if the map wasn't handled?

No, assert is wrong -- these crc errors in the incremental maps are somewhat normal -- they can occur during an upgrade, for example. Check the log that Xiaoxi sent here: https://tracker.ceph.com/issues/46443#note-11 there you can see lots of incrementals that get crc error, then a full is requested and received correctly. The _committed_osd_maps crash only happens when the crc error is unluckily the first map in the message.

> Could the claims in #35218 not be correct?

They make sense to me. consume_map is indeed probably expensive, and useless if we're re-consuming the current map.